// mini-extract-plugin/dist/models/dependency.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Dependency_1 = __importDefault(require("webpack/lib/Dependency"));
const lodash_capitalize_1 = __importDefault(require("lodash.capitalize"));
const util_1 = require("../lib/util");
const TypedWebpackDep = Dependency_1.default;
class Dependency extends TypedWebpackDep {
    constructor({ identifier, content, moduleType, miniExtractType }, context, identifierIndex) {
        super();
        this.moduleType = moduleType;
        this.miniExtractType = miniExtractType;
        this.identifier = identifier;
        this.identifierIndex = identifierIndex;
        this.context = context;
        this.content = content;
    }
    getResourceIdentifier() {
        return `${this.miniExtractType}-module-${this.identifier}-${this.identifierIndex}`;
    }
}
exports.Dependency = Dependency;
function subclass({ type } = {}) {
    class DependencySubclass extends Dependency {
        getResourceIdentifier() {
            return `${type}-module-${this.identifier}-${this.identifierIndex}`;
        }
    }
    const className = `${lodash_capitalize_1.default(type)}${Dependency.name}`;
    util_1.renameClass(DependencySubclass, className);
    return DependencySubclass;
}
exports.subclass = subclass;
